webrender/src/device/gfx/descriptor.rs

```diff
@@ -192,19 +201,97 @@ impl<B: hal::Backend> DescriptorPools<B> {
+                            self.set,
+                        );
+                        // In lot of cases when we need extra pools, we will need an enormous amount of descriptors (above 4000).
+                        // Because of this we double the size of each new pool compared to the previous one.
+                        let mul = 2_usize.pow(self.cache_clip_pool_idx as u32).min(4096);
+                        Self::increase_range_count(&mut cache_clip_layout_range, self.desc_per_pool * mul);
+                        self.cache_clip_pool.push(DescPool::new(
+                            device,
+                            self.desc_per_pool * mul,
+                            cache_clip_layout_range,
+                            cache_clip_layout,
+                        ));
+                        assert!(self.get_pool_mut(shader_kind).next());
+                    }
+                },
+                 _ => {
+                    self.default_pool_idx += 1;
+                    if self.default_pool_idx < self.default_pool.len() {
+                        assert!(self.get_pool_mut(shader_kind).next());
+                    } else {
+                        let (default_layout, mut default_layout_range) = Self::get_layout_and_range(
+                            pipeline_requirements
+                                .get("brush_solid")
+                                .expect("brush_solid missing"),
+                            self.set,
+                        );
+                        let mul = 2_usize.pow(self.default_pool_idx as u32).min(4096);
+                        Self::increase_range_count(&mut default_layout_range, self.desc_per_pool  * mul);
+                        self.default_pool.push(DescPool::new(
+                            device,
+                            self.desc_per_pool * mul,
+                            default_layout_range,
+                            default_layout,
+                        ));
+                        assert!(self.get_pool_mut(shader_kind).next());
+                    }
+                },
+            }
+        }
     }
 
-    pub(super) fn reset(&mut self) {
+    pub(super) fn reset(&mut self, device: &B::Device) {
         self.debug_pool.reset();
-        self.cache_clip_pool.reset();
-        self.default_pool.reset();
+
+        // Free descritor pools which were not used in the previous draw
+        while self.cache_clip_pool_idx < self.cache_clip_pool.len() - 1 {
+            let pool = self.cache_clip_pool.pop()
+                .expect("No cache clip pool found");
+            pool.deinit(device);
+        }
+        while self.default_pool_idx < self.default_pool.len() - 1 {
+            let pool = self.default_pool.pop()
+                .expect("No default pool found");;
+            pool.deinit(device);
+        }
+
+        for pool in self.cache_clip_pool.iter_mut() {
+            pool.reset()
+        }
```

Comment:
Maybe `self.cache_clip_pool.iter_mut().map(|pool| pool.reset());` ? And similarly for the rest of the `for`s.

Reply:
It would only work if we collect the result of that expression maybe into a `Vec`, because `map` is performed lazily, and I think it's better that using a for loop.